--- HTML/script.js.orig	2025-05-20 15:57:21.700332003 +0200
+++ HTML/script.js	2025-05-20 15:57:21.700332003 +0200
@@ -38,7 +38,8 @@
 var range_update_time = null;
 let updateInProgress = false;
 let activeTileLayer = undefined;
-var
+var hover_enabled_track = false,
+    select_enabled_track = false,
     marker_tracks = new Set(),
     marker_fireworks = [];
 
@@ -100,7 +101,10 @@
         labels_declutter: true,
         eri: true,
         loadURL: true,
-        map_opacity: 0.5
+        map_opacity: 0.5,
+        show_track_on_hover: false,
+        show_track_on_select: false,
+        shiptable_columns: ["shipname", "mmsi", "imo", "callsign", "shipclass", "lat", "lon", "last_signal", "level", "distance", "bearing", "speed", "repeat", "ppm", "status"]
     };
 }
 
@@ -114,6 +118,7 @@
     let android = settings.android;
     let darkmode = settings.dark_mode;
     restoreDefaultSettings();
+
     settings.android = android;
     if (isAndroid()) settings.dark_mode = darkmode;
 
@@ -179,16 +184,33 @@
 getDistanceUnit = () => (settings.metric === "DEFAULT" ? "nmi" : settings.metric === "SI" ? "km" : "mi");
 getSpeedVal = (c) => (settings.metric === "DEFAULT" ? Number(c).toFixed(1) : settings.metric === "SI" ? Number(c * 1.852).toFixed(1) : Number(c * 1.151).toFixed(1));
 getSpeedUnit = () => (settings.metric === "DEFAULT" ? "kts" : settings.metric === "SI" ? "km/h" : "mph");
+getShipDimension = (ship) => ship.to_bow != null && ship.to_stern != null && ship.to_port != null && ship.to_starboard != null
+    ? getDimVal(ship.to_bow + ship.to_stern) + " " + getDimUnit() + " x " + getDimVal(ship.to_port + ship.to_starboard) + " " + getDimUnit()
+    : null
 
 getLatValFormat = (ship) => (ship.approx ? "<i>" : "") + (settings.latlon_in_dms ? decimalToDMS(ship.lat, true) : Number(ship.lat).toFixed(5)) + (ship.approx ? "</i>" : "");
 getLonValFormat = (ship) => (ship.approx ? "<i>" : "") + (settings.latlon_in_dms ? decimalToDMS(ship.lon, false) : Number(ship.lon).toFixed(5)) + (ship.approx ? "</i>" : "");
 
 getEtaVal = (ship) => ("0" + ship.eta_month).slice(-2) + "-" + ("0" + ship.eta_day).slice(-2) + " " + ("0" + ship.eta_hour).slice(-2) + ":" + ("0" + ship.eta_minute).slice(-2);
-getDeltaTimeVal = (s) => (s < 60 ? s + "s" : s < 60 * 60 ? Math.floor(s / 60) + "m " + (s % 60) + "s" : Math.floor(s / 3600) + "h " + Math.floor((s % 3600) / 60) + "m");
 getShipName = (ship) => ship.shipname;
 getCallSign = (ship) => ship.callsign;
 includeShip = (ship) => true;
 
+const getDeltaTimeVal = (s) => {
+    const days = Math.floor(s / (24 * 3600));
+    const hours = Math.floor((s % (24 * 3600)) / 3600);
+    const minutes = Math.floor((s % 3600) / 60);
+    const seconds = s % 60;
+
+    let result = '';
+    if (days > 0) result += `${days}d `;
+    if (hours > 0 || days > 0) result += `${hours}h `;
+    if (minutes > 0 || hours > 0 || days > 0) result += `${minutes}m `;
+    if (seconds > 0 || (days === 0 && hours === 0 && minutes === 0)) result += `${seconds}s`;
+
+    return result.trim();
+};
+
 const notificationContainer = document.getElementById("notification-container");
 
 // https://stackoverflow.com/questions/51805395/navigator-clipboard-is-undefined
@@ -297,7 +319,7 @@
         c = settings.shiphover_color;
         w = w + 2;
     }
-    if (feature.mmsi == card_mmsi) {
+    else if (feature.mmsi == card_mmsi) {
         c = settings.shipselection_color;
         w = w + 2;
     }
@@ -328,11 +350,17 @@
     });
 };
 
+function decodeHTMLEntities(text) {
+    const textArea = document.createElement('textarea');
+    textArea.innerHTML = text;
+    return textArea.value;
+}
+
 var labelStyle = function (feature) {
     const font = settings.tooltipLabelFontSize + "px Arial";
     return new ol.style.Style({
         text: new ol.style.Text({
-            text: feature.ship.shipname || feature.ship.mmsi.toString(),
+            text: decodeHTMLEntities(feature.ship.shipname || feature.ship.mmsi.toString()),
             overflow: true,
             offsetY: 25,
             offsetX: 25,
@@ -354,7 +382,7 @@
             radius: 20,
             stroke: new ol.style.Stroke({
                 color: settings.shiphover_color,
-                width: 5
+                width: 8
             })
         })
     });
@@ -366,7 +394,7 @@
             radius: 15,
             stroke: new ol.style.Stroke({
                 color: settings.shipselection_color,
-                width: 5
+                width: 8
             })
         })
     });
@@ -1270,7 +1298,7 @@
 
         closeDialog();
         closeSettings();
-        showShipcard(feature.ship.mmsi);
+        showShipcard(feature.ship.mmsi, pixel);
     }
     else {
         clickTimeout = setTimeout(function () {
@@ -1683,8 +1711,8 @@
                 cell2.innerText = shipName;
 
                 var cell3 = row.insertCell(2);
-                cell3.innerHTML = ship.distance ? getDistanceVal(ship.distance) : "";
-                cell3.title = ship.distance != null ? getDistanceVal(ship.distance) + " " + getDistanceUnit() : "";
+                cell3.innerHTML = ship.distance ? (getDistanceVal(ship.distance) + (ship.repeat > 0 ? " (R)" : "")) : null
+                cell3.title = ship.distance != null ? getDistanceVal(ship.distance) + " " + getDistanceUnit() + (ship.repeat > 0 ? " (R)" : "") : "";
 
                 var cell4 = row.insertCell(3);
                 cell4.innerHTML = ship.speed != null ? getSpeedVal(ship.speed) : "";
@@ -2361,9 +2389,7 @@
         "shiptype",
         "mmsi_type",
         "shipclass",
-        "validated",
         "msg_type",
-        "channels",
         "country",
         "status",
         "draught",
@@ -2376,6 +2402,10 @@
         "shipname",
         "destination",
         "last_signal",
+        "flags",
+        "validated",
+        "channels"
+
     ];
 
     shipsDB = {};
@@ -2384,6 +2414,23 @@
     ships.values.forEach((v) => {
         const s = Object.fromEntries(keys.map((k, i) => [k, v[i]]));
 
+        const flags = s.flags;
+        s.validated2 = (flags & 3) == 2 ? -1 : flags & 3;
+
+
+        s.repeat = (flags >> 2) & 3;
+        s.virtual_aid = (flags >> 4) & 1;
+        s.approximate = (flags >> 5) & 1;
+        s.channels2 = (flags >> 6) & 0b1111;
+
+        // Check for discrepancies and show error
+        if (s.validated !== s.validated2) {
+            console.log(`Mismatch in validated: ${s.validated} != ${s.validated2}`);
+        }
+        if (s.channels !== s.channels2) {
+            console.log(`Mismatch in channels: ${s.channels} != ${s.channels2}`);
+        }
+
         if (includeShip(s)) {
             s.shipname = sanitizeString(s.shipname);
             s.callsign = sanitizeString(s.callsign);
@@ -2520,16 +2567,21 @@
     if (d.chart.data.datasets.length > 1) start = 1;
 
     var c = 0;
+
     for (a = 0; a < b.length; a++) {
         if (b[a].x != 0) {
             for (i = start; i < d.chart.data.datasets.length; i++) {
-                c += d.chart.data.datasets[i].data[a].y;
+                if (!d.chart.getDatasetMeta(i).hidden) {
+                    c += d.chart.data.datasets[i].data[a].y;
+                }
             }
         }
     }
+
     return c / (b.length - 1);
 }
 
+
 const graph_annotation = {
     type: "line",
     borderColor: "rgba(12,118,170)",
@@ -2999,6 +3051,15 @@
                 (!e[i].classList.contains("shipcard-max-only") && !e[i].classList.contains("shipcard-row-selected"))
             )
                 e[i].classList.toggle("shipcard-row-selected");
+
+            const aside = document.getElementById("shipcard");
+            if (aside.style.top && aside.getBoundingClientRect().bottom > window.innerHeight) {
+
+                if (card_mmsi in shipsDB) {
+                    let pixel = map.getPixelFromCoordinate(ol.proj.fromLonLat([shipsDB[card_mmsi].raw.lon, shipsDB[card_mmsi].raw.lat]));
+                    positionAside(pixel, aside);
+                }
+            }
         }
     } else {
         for (let i = 0; i < e.length; i++) {
@@ -3319,6 +3380,8 @@
         (data.group_mask != null && getStringfromGroup(data.group_mask).includes(query));
 }
 
+let tableFirstTime = true;
+
 async function updateShipTable() {
     const ok = await fetchShips();
     if (!ok) return;
@@ -3350,6 +3413,17 @@
                     },
                 },
                 { title: "MMSI", field: "mmsi", sorter: "number" },
+                { title: "IMO", field: "imo", sorter: "number" },
+                { title: "Dest", field: "destination", sorter: "string" },
+                {
+                    title: "ETA",
+                    field: "eta",
+                    sorter: "string",
+                    formatter: function (cell) {
+                        const ship = cell.getRow().getData();
+                        return ship.eta_month != null && ship.eta_hour != null && ship.eta_day != null && ship.eta_minute != null ? getEtaVal(ship) : null;
+                    },
+                },
                 {
                     title: "Callsign",
                     field: "callsign",
@@ -3360,6 +3434,24 @@
                     },
                 },
                 {
+                    title: "Flag",
+                    field: "country",
+                    sorter: "string",
+                    formatter: function (cell) {
+                        const ship = cell.getRow().getData();
+                        return ship != null ? getCountryName(ship.country) : "";
+                    },
+                },
+                {
+                    title: "Status",
+                    field: "status",
+                    sorter: "string",
+                    formatter: function (cell) {
+                        const ship = cell.getRow().getData();
+                        return ship != null ? getStatusVal(ship) : "";
+                    },
+                },
+                {
                     title: "Type",
                     field: "shipclass",
                     sorter: "number",
@@ -3369,6 +3461,33 @@
                     },
                 },
                 {
+                    title: "Class",
+                    field: "class",
+                    sorter: "string",
+                    formatter: function (cell) {
+                        const ship = cell.getRow().getData();
+                        return getTypeVal(ship);
+                    },
+                },
+                {
+                    title: "Dim",
+                    field: "dimension",
+                    sorter: "number",
+                    formatter: function (cell) {
+                        const ship = cell.getRow().getData();
+                        return ship ? getShipDimension(ship) || "" : "";
+                    },
+                },
+                {
+                    title: "Draught",
+                    field: "draught",
+                    sorter: "number",
+                    formatter: function (cell) {
+                        const ship = cell.getRow().getData();
+                        return ship.draught ? getDimVal(ship.draught) + " " + getDimUnit() : null;
+                    },
+                },
+                {
                     title: "Last",
                     field: "last_signal",
                     sorter: "number",
@@ -3377,7 +3496,7 @@
                         return value != null ? getDeltaTimeVal(value) : "";
                     },
                 },
-                { title: "Msgs", field: "count", sorter: "number" },
+                { title: "Count", field: "count", sorter: "number" },
                 {
                     title: "PPM",
                     field: "ppm",
@@ -3401,8 +3520,8 @@
                     field: "distance",
                     sorter: "number",
                     formatter: function (cell) {
-                        const value = cell.getValue();
-                        return value != null ? getDistanceVal(value) : "";
+                        const ship = cell.getRow().getData();
+                        return ship != null ? getDistanceVal(ship.distance) + (ship.repeat > 0 ? " (R)" : "") : "";
                     },
                 },
                 {
@@ -3454,6 +3573,15 @@
                     },
                 },
                 {
+                    title: "Repeated",
+                    field: "repeat",
+                    sorter: "string",
+                    formatter: function (cell) {
+                        const value = cell.getValue();
+                        return value == 1 ? "Yes" : "No";
+                    },
+                },
+                {
                     title: "Ch",
                     field: "channels",
                     sorter: "number",
@@ -3463,6 +3591,42 @@
                     },
                 },
                 {
+                    title: "Msgs",
+                    field: "msg_type",
+                    sorter: "number",
+                    formatter: function (cell) {
+                        const value = cell.getValue();
+                        return getStringfromMsgType(value);
+                    },
+                },
+                {
+                    title: "MSG6",
+                    field: "MSG6",
+                    sorter: "number",
+                    formatter: function (cell) {
+                        const value = cell.getValue();
+                        return value & (1 << 6) ? "Yes" : "No";
+                    },
+                },
+                {
+                    title: "MSG8",
+                    field: "MSG8",
+                    sorter: "number",
+                    formatter: function (cell) {
+                        const value = cell.getValue();
+                        return value & (1 << 8) ? "Yes" : "No";
+                    },
+                },
+                {
+                    title: "MSG27",
+                    field: "MSG27",
+                    sorter: "number",
+                    formatter: function (cell) {
+                        const value = cell.getValue();
+                        return value & (1 << 27) ? "Yes" : "No";
+                    },
+                },
+                {
                     title: "Src",
                     field: "group_mask",
                     sorter: "number",
@@ -3479,6 +3643,14 @@
         table.on("rowClick", function (e, row) {
             tableRowClick(row.getData().mmsi);
         });
+        table.on("tableBuilt", function () {
+            if (tableFirstTime) {
+                setShipTableColumns();
+                tableFirstTime = false;
+            }
+
+            populateShipTableColumnVisibilityMenu();
+        });
     } else {
         if (data.length == 0) {
             table.clearData();
@@ -3494,6 +3666,92 @@
     }
 }
 
+function populateShipTableColumnVisibilityMenu() {
+    const shipTableColumnVisibilityMenu = document.getElementById("shipTableColumnVisibilityMenu");
+    shipTableColumnVisibilityMenu.innerHTML = "";
+
+    table.getColumns().forEach((column) => {
+        const checkboxId = `checkbox-${column.getField()}`;
+        const checkboxLabel = column.getDefinition().title;
+
+        const checkboxItem = document.createElement("div");
+        checkboxItem.classList.add("ship-table-column-dropdown-item");
+
+        const checkbox = document.createElement("input");
+        checkbox.type = "checkbox";
+        checkbox.id = checkboxId;
+        checkbox.value = column.getField();
+        checkbox.checked = column.isVisible();
+        checkbox.addEventListener("change", updateShipTableColumnVisibility);
+
+        const label = document.createElement("label");
+        label.setAttribute("for", checkboxId);
+        label.textContent = checkboxLabel;
+
+        checkboxItem.appendChild(checkbox);
+        checkboxItem.appendChild(label);
+
+        shipTableColumnVisibilityMenu.appendChild(checkboxItem);
+    });
+}
+
+function updateShipTableColumnVisibility() {
+    const checkboxes = document.querySelectorAll("#shipTableColumnVisibilityMenu input[type='checkbox']");
+
+    settings.shiptable_columns = Array.from(checkboxes)
+        .filter((checkbox) => checkbox.checked)
+        .map((checkbox) => checkbox.value);
+
+    setShipTableColumns();
+}
+
+function resetShipTableColumns() {
+    settings.shiptable_columns = ["shipname", "mmsi", "imo", "callsign", "shipclass", "lat", "lon", "last_signal", "level", "distance", "bearing", "speed", "repeat", "ppm", "status"];
+    setShipTableColumns();
+}
+
+function setShipTableColumns() {
+
+    saveSettings();
+
+    const allColumns = table.getColumns();
+
+    const updatedColumns = allColumns.map(column => {
+        const field = column.getField();
+        return {
+            ...column.getDefinition(),
+            visible: settings.shiptable_columns.includes(field)
+        };
+    });
+
+    table.setColumns(updatedColumns);
+    populateShipTableColumnVisibilityMenu();
+}
+
+document.getElementById("shipTableColumnDropdownToggle").addEventListener("click", populateShipTableColumnVisibilityMenu);
+const dropdownToggle = document.getElementById('shipTableColumnDropdownToggle');
+const dropdownMenu = document.getElementById('shipTableColumnVisibilityMenu');
+
+function toggleDropdown() {
+    dropdownMenu.style.display = dropdownMenu.style.display === 'block' ? 'none' : 'block';
+}
+
+dropdownToggle.addEventListener('click', function (event) {
+    event.stopPropagation();
+    toggleDropdown();
+});
+
+document.addEventListener('click', function (event) {
+    if (!dropdownMenu.contains(event.target) && event.target !== dropdownToggle) {
+        dropdownMenu.style.display = 'none';
+    }
+});
+
+const resetButton = document.getElementById('shipTableColumnReset');
+resetButton.addEventListener('click', function (event) {
+    resetShipTableColumns();
+});
+
 function getTooltipContent(ship) {
     return '<div>' + getFlagStyled(ship.country, "padding: 0px; margin: 0px; margin-right: 10px; margin-left: 3px; box-shadow: 1px 1px 2px rgba(0, 0, 0, 0.2); font-size: 26px;") + `</div><div><div><b>${getShipName(ship) || ship.mmsi}</b> at <b>${getSpeedVal(ship.speed)} ${getSpeedUnit()}</b></div><div>Received <b>${getDeltaTimeVal(ship.last_signal)}</b> ago</div></div>`;
 }
@@ -3557,12 +3815,17 @@
 
 
 const stopHover = function () {
+
+    if (!hoverMMSI) return;
+
     hover_info.style.visibility = 'hidden';
 
     if (hoverMMSI in shapeFeatures) {
         shapeFeatures[hoverMMSI].changed();
     }
 
+    if (hover_enabled_track) hideTrack(hoverMMSI);
+
     hoverMMSI = undefined;
     updateHoverMarker();
     trackLayer.changed();
@@ -3627,6 +3890,10 @@
         }
         updateHoverMarker();
         trackLayer.changed();
+        if (settings.show_track_on_hover) {
+            hover_enabled_track = !trackIsShown(hoverMMSI);
+            if (hover_enabled_track) showTrack(mmsi);
+        }
         if (mmsi in shapeFeatures) {
             shapeFeatures[mmsi].changed();
         }
@@ -3852,6 +4119,22 @@
     }
 }
 
+async function showTrack(m) {
+    if (!marker_tracks.has(Number(m))) {
+        toggleTrack(m);
+    }
+}
+
+async function hideTrack(m) {
+    if (marker_tracks.has(Number(m))) {
+        toggleTrack(m);
+    }
+}
+
+function trackIsShown(m) {
+    return marker_tracks.has(Number(m));
+}
+
 function pinStation() {
     pinVessel("STATION");
 }
@@ -3872,6 +4155,7 @@
 
 async function showAllTracks() {
     show_all_tracks = true;
+    select_enabled_track = hover_enabled_track = false;
     await fetchTracks();
     redrawMap();
     updateShipcardTrackOption()
@@ -4039,12 +4323,9 @@
     document.getElementById("shipcard_lon").innerHTML = ship.lon ? getLonValFormat(ship) : null;
 
     document.getElementById("shipcard_speed").innerHTML = ship.speed ? getSpeedVal(ship.speed) + " " + getSpeedUnit() : null;
-    document.getElementById("shipcard_distance").innerHTML = ship.distance ? getDistanceVal(ship.distance) + " " + getDistanceUnit() : null;
+    document.getElementById("shipcard_distance").innerHTML = ship.distance ? (getDistanceVal(ship.distance) + " " + getDistanceUnit() + (ship.repeat > 0 ? " (R)" : "")) : null;
     document.getElementById("shipcard_draught").innerHTML = ship.draught ? getDimVal(ship.draught) + " " + getDimUnit() : null;
-    document.getElementById("shipcard_dimension").innerHTML =
-        ship.to_bow != null && ship.to_stern != null && ship.to_port != null && ship.to_starboard != null
-            ? getDimVal(ship.to_bow + ship.to_stern) + " " + getDimUnit() + " x " + getDimVal(ship.to_port + ship.to_starboard) + " " + getDimUnit()
-            : null;
+    document.getElementById("shipcard_dimension").innerHTML = getShipDimension(ship);
 
     document.getElementById("shipcard_track").innerText = marker_tracks.has(Number(card_mmsi)) ? "Hide Track" : "Show Track";
 }
@@ -4128,16 +4409,130 @@
     }
 }
 
-function showShipcard(m) {
+function adjustMapForShipcard(ship, pixel) {
+    if (ship && ship.lon && ship.lat) {
+        let view = map.getView();
+        // Use the provided pixel if defined, otherwise calculate it
+        if (!pixel) {
+            let shipCoords = ol.proj.fromLonLat([ship.lon, ship.lat]);
+            pixel = map.getPixelFromCoordinate(shipCoords);
+        }
+
+        let shipcard = document.getElementById("shipcard");
+        let shipcardRect = shipcard.getBoundingClientRect();
+        let mapElement = map.getTargetElement();
+        let mapRect = mapElement.getBoundingClientRect();
+
+        let isUnderShipcard = (
+            pixel[0] + mapRect.left >= shipcardRect.left &&
+            pixel[0] + mapRect.left <= shipcardRect.right &&
+            pixel[1] + mapRect.top >= shipcardRect.top &&
+            pixel[1] + mapRect.top <= shipcardRect.bottom
+        );
+
+        if (isUnderShipcard) {
+            let newPixel = [...pixel];
+            let margin = 10; // Margin in pixels
+
+            if (shipcardRect.bottom + margin + 20 <= mapRect.bottom) { // 20 is an approximate marker height
+                newPixel[1] = shipcardRect.bottom + margin;
+            }
+            else if (shipcardRect.right + margin + 20 <= mapRect.right) { // 20 is an approximate marker width
+                newPixel[0] = shipcardRect.right + margin;
+            }
+            newPixel = pixel;
+
+
+            console.log("Original pixel:", pixel);
+            console.log("New pixel:", newPixel);
+
+            let currentCenter = view.getCenter();
+            let newCenter = map.getCoordinateFromPixel(newPixel);
+            let centerDiff = [newCenter[0] - currentCenter[0], newCenter[1] - currentCenter[1]];
+
+            stopHover();
+            view.animate({
+                center: [currentCenter[0] + centerDiff[0], currentCenter[1] + centerDiff[1]],
+                duration: 1000
+            });
+        }
+    }
+}
+
+function positionAside(pixel, aside) {
+
+    stopHover();
+    aside.style.left = "";
+    aside.style.top = "";
+
+    if (pixel) {
+        const margin = 35;
+
+        let marginRight = document.getElementById("tableside").classList.contains("active") ? 592 : 30;
+
+        const mapSize = map.getSize();
+        const shipCardRect = aside.getBoundingClientRect();
+        const shipCardWidth = shipCardRect.width;
+        const shipCardHeight = shipCardRect.height;
+
+        const rightSpace = mapSize[0] - (pixel[0] + shipCardWidth + margin + marginRight);
+        const leftSpace = pixel[0] - (shipCardWidth + margin);
+
+        if ((rightSpace > 0 || leftSpace > 0) && mapSize[1] > shipCardHeight + 2 * margin) {
+
+            let topPosition = pixel[1] - (shipCardHeight / 2);
+            topPosition = Math.max(margin, Math.min(mapSize[1] - shipCardHeight - margin, topPosition));
+
+            aside.style.top = `${topPosition}px`;
+
+            if (rightSpace >= 0) {
+                aside.style.left = `${pixel[0] + margin}px`;
+            } else if (leftSpace >= 0) {
+                aside.style.left = `${pixel[0] - shipCardWidth - margin}px`;
+            } else {
+                aside.style.left = `${(mapSize[0] - shipCardWidth) / 2}px`;
+            }
+        }
+    }
+    adjustMapForShipcard(ship, pixel);
+}
+
+function showShipcard(m, pixel = undefined) {
     const aside = document.getElementById("shipcard");
     const visible = shipcardVisible();
 
     ship = m in shipsDB ? shipsDB[m].raw : null;
     ship_old = card_mmsi in shipsDB ? shipsDB[card_mmsi].raw : null;
 
+    if (select_enabled_track && (card_mmsi != m || m == null)) {
+        if (trackIsShown(card_mmsi)) {
+            hideTrack(card_mmsi);
+            select_enabled_track = false;
+        }
+    }
+
+    if (settings.show_track_on_select && m != null) {
+        if (!trackIsShown(m)) {
+            select_enabled_track = true;
+            showTrack(m);
+        }
+        else {
+            if (hoverMMSI == m && hover_enabled_track) {
+                hover_enabled_track = false;
+                select_enabled_track = true;
+            }
+        }
+    }
+
     if (m != null && !visible) {
         if (measurecardVisible()) toggleMeasurecard();
         aside.classList.toggle("visible");
+
+        if (settings.show_track_on_select && !trackIsShown(m)) {
+            select_enabled_track = true;
+            showTrack(m);
+        }
+
     } else if (visible && m == null) {
         aside.classList.toggle("visible");
     }
@@ -4150,18 +4545,11 @@
     if (shipcardVisible()) {
         populateShipcard();
 
-        if (ship && ship.lon && ship.lat) {
-            let shipCoords = ol.proj.fromLonLat([ship.lon, ship.lat]);
-
-            let view = map.getView();
-            let extent = view.calculateExtent(map.getSize());
-
-            if (!ol.extent.containsCoordinate(extent, shipCoords)) {
-                view.setCenter(shipCoords);
-                //view.setZoom(view.getMaxZoom());
-            }
+        if (isShipcardMax()) {
+            toggleShipcardSize();
         }
         if (!visible) shipcardMinIfMaxonMobile();
+        positionAside(pixel, aside);
     }
 }
 
@@ -4542,6 +4930,10 @@
     document.getElementById("settings_tooltipLabelColorDark").value = settings.tooltipLabelColorDark;
     document.getElementById("settings_tooltipLabelShadowColorDark").value = settings.tooltipLabelShadowColorDark;
     document.getElementById("settings_table_shiptype_use_icon").checked = settings.table_shiptype_use_icon;
+    document.getElementById("settings_show_track_on_hover").checked = settings.show_track_on_hover;
+    document.getElementById("settings_show_track_on_select").checked = settings.show_track_on_select;
+
+
 }
 
 function activateTab(b, a) {
